groups/reconcile.go

```diff
@@ -181,69 +178,172 @@ func main() {
+// interfaces.
+type Reconciler struct {
+	AdminSvc AdminService
+	GroupSvc GroupService
+}
+
+func NewReconciler(ctx context.Context, clientOption option.ClientOption) (*Reconciler, error) {
+	as, err := NewAdminService(ctx, clientOption)
+	if err != nil {
+		return nil, err
+	}
+
+	gs, err := NewGroupService(ctx, clientOption)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Reconciler{AdminSvc: as, GroupSvc: gs}, nil
+}
+
+func (r *Reconciler) ReconcileGroups(groups []GoogleGroup) error {
+	for _, g := range groups {
 		if g.EmailId == "" {
-			log.Fatal(fmt.Sprintf("Group has no email-id: %#v", g))
+			return fmt.Errorf("group has no email-id: %#v", g)
 		}
 
-		err = createOrUpdateGroupIfNecessary(adminService, g.EmailId, g.Name, g.Description)
+		err := r.AdminSvc.CreateOrUpdateGroupIfNescessary(g)
 		if err != nil {
-			log.Fatal(err)
+			return err
 		}
-		err = updateGroupSettings(groupsSettingsService, g.EmailId, g.Settings)
+		err = r.GroupSvc.UpdateGroupSettings(g)
 		if err != nil {
-			log.Fatal(err)
+			return err
 		}
-		err = addOrUpdateMembersToGroup(adminService, g.EmailId, g.Owners, "OWNER")
+		err = r.AdminSvc.AddOrUpdateGroupMembers(g, OwnerRole, g.Owners)
 		if err != nil {
-			log.Fatal(err)
+			return err
 		}
-		err = addOrUpdateMembersToGroup(adminService, g.EmailId, g.Managers, "MANAGER")
+		err = r.AdminSvc.AddOrUpdateGroupMembers(g, ManagerRole, g.Managers)
 		if err != nil {
-			log.Fatal(err)
+			return err
 		}
-		err = addOrUpdateMembersToGroup(adminService, g.EmailId, g.Members, "MEMBER")
+		err = r.AdminSvc.AddOrUpdateGroupMembers(g, MemberRole, g.Members)
 		if err != nil {
 			log.Println(err)
 		}
+
 		if g.Settings["ReconcileMembers"] == "true" {
 			members := append(g.Owners, g.Managers...)
 			members = append(members, g.Members...)
-			err = removeMembersFromGroup(adminService, g.EmailId, members)
+			err = r.AdminSvc.RemoveMembersFromGroup(g, members)
 			if err != nil {
-				log.Fatal(err)
+				return err
 			}
 		} else {
 			members := append(g.Owners, g.Managers...)
-			err = removeOwnerOrManagersGroup(adminService, g.EmailId, members)
+			err = r.AdminSvc.RemoveOwnerOrManagersFromGroup(g, members)
 			if err != nil {
-				log.Fatal(err)
+				return err
 			}
 		}
 	}
-	err = deleteGroupsIfNecessary(adminService)
+	err := r.AdminSvc.DeleteGroupsIfNecessary()
 	if err != nil {
-		log.Fatal(err)
+		return err
+	}
+
+	return nil
+}
+
+func (r *Reconciler) printGroupMembersAndSettings() error {
+	asClient := r.AdminSvc.GetClient()
+	gsClient := r.GroupSvc.GetClient()
+
+	g, err := asClient.ListGroups()
+	if err != nil {
+		return fmt.Errorf("unable to retrieve users in domain: %v", err)
+	}
+
+	if len(g.Groups) == 0 {
+		log.Println("No groups found.")
+		return nil
+	}
+
+	var groupsConfig GroupsConfig
+	for _, g := range g.Groups {
+		group := GoogleGroup{
+			EmailId:     g.Email,
+			Name:        g.Name,
+			Description: g.Description,
+		}
+		g2, err := gsClient.Get(g.Email)
```

Comment:
e.g. this could probably be a `GroupService` function so we don't need the client